### PR TITLE
Validate dashboard name length and dashboardTab length

### PR DESCRIPTION
**Pull request by chases2**
Fixes https://github.com/GoogleCloudPlatform/testgrid/issues/98

- Tests added to check for empty dashboards and return an error in that case
- Tests added to check for a short entity length: 0-2 length entities are not valid

### CODE REVIEW

File: config/config.go [hunk 1]
```diff
@@ -211,8 +212,9 @@ func validateReferencesExist(c *configpb.Configuration) error {
 // TestGrid file prefix, post-normalization.
 func validateName(s string) error {
 	name := normalize(s)
-	if name == "" {
-		return errors.New("normalized name can't be empty")
+
+	if len(name) < 3 {
+		return errors.New("names must contain at least 3 alphanumeric characters")
```
Comment (fejta) [inline config/config.go:217]: Shouldn't there also be a maximum length?

File: config/config.go [hunk 2]
```diff
@@ -443,6 +445,13 @@ func Validate(c *configpb.Configuration) error {
 		return multierror.Append(mErr, MissingFieldError{"Dashboards"})
 	}
 
+	// Each Dashboard must contain at least 1 Tab to do anything
+	for _, dashboard := range c.GetDashboards() {
+		if len(dashboard.DashboardTab) == 0 {
```
Comment (fejta) [inline config/config.go:450]: `dashboard.DashboardTab != ""` is the (faster and) more idiomatic way of writing this.

**Review by fejta — APPROVED** [review]
/hold

File: config/config.go (hunk 1 above)
Comment (chases2) [reply config/config.go:217]: Maybe. What should it be?

File: config/config.go (hunk 2 above)
Comment (chases2) [reply config/config.go:450]: dashboard.DashboardTab is a []*DashboardTab, not a string

File: config/config.go (hunk 1 above)
Comment (fejta) [reply config/config.go:217]: I'd just pick a longish number and ensure it doesn't exceed this (such as 512, 1024 characters). Doesn't particularly matter.

File: config/config.go [hunk 3]
```diff
@@ -211,8 +212,13 @@ func validateReferencesExist(c *configpb.Configuration) error {
...
 func validateName(s string) error {
 	name := normalize(s)
-	if name == "" {
-		return errors.New("normalized name can't be empty")
+
+	if len(name) < 3 {
+		return errors.New("names must contain at least 3 alphanumeric characters")
+	}
+
+	if len(name) > 2048 {
```
Comment (fejta) [inline config/config.go:220]: Add unit test?

I would make these consts for a more robust unit test, aka:

```golang
const small = 3, big = 4

cases := []struct{
  name string
  n int
  pass bool
}{
  {
    name: "too small fails",
    n: small-1,
    err: true,
  },
  {
    name: "small pass",
    n: small,
    err: true,
   },
   // same for big and big+1
}
```

also

```golang
var sb strings.Builder
for i := 0; i<tc.n;i++ { 
  sb.WriteRune('a')
}
name := sb.String()
```

Comment (fejta) [reply config/config.go:220]: (AKA that way we can change big from 2048 to 1024 or whatever without updating the unit test)
(Also fine to just use the 2048 literal in the unit test)

**Review by fejta — APPROVED** [review]
/hold
would like to see missing coverage added in either this PR or next one

File: config/config.go (hunk 3 above)
Comment (chases2) [reply config/config.go]: This feels like overkill, but it does give me an opportunity to make existing tests a bit more descriptive.

File: config/config_test.go [hunk 4]
```diff
@@ -363,37 +364,70 @@ func TestValidateReferencesExist(t *testing.T) {
 	}
 }
 
-func TestUpdate_validateNames(t *testing.T) {
+func TestValidateName(t *testing.T) {
+	stringOfLength := func(length int) string {
```
Comment (fejta) [inline config/config_test.go:368]: IMO this function doesn't need to exist as it is only called once right now.

In the future when someone is adding a new test to this list and/or debugging a failure that isn't the "Names can't be too long" test, this adds unnecessary complexity for that person, reducing their performance.

**Review by fejta — APPROVED** [review]
/hold

**Pull request merged**